KAFKA-13258/13259/13260: Fix error response generation

mimaison opened the pull request:
AlterClientQuotas, DescribeProducers and FindCoordinator have issues when building error responses. This can lead to brokers returning responses without errors even when some have happened.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

mimaison committed:
MINOR: Fix error response generation

AlterClientQuotas, DescribeProducers and FindCoordinator have issues when building error responses. This can lead to brokers returning responses without errors even when some have happened.

mimaison commented:
This affects 2.8.0 too so part of this PR will need to be backported.

dajac commented:
@mimaison Is it a regression? It would be good to file a bug for it in order to have a trace about it. I was about to cut a RC for 2.8.1 but I'll wait on this one.

mimaison commented:
Yes it's partly a regression:
- AlterClientQuotas worked correctlyup to 2.7.0 and is broken since 2.8.0.
- FindCoordinator worked correctly up to 2.8.0 and is broken since 3.0.0

DescribeProducers is broken since releasing in 2.8.0.

It almost feels like we should create a ticket (and PR) for each of them

mimaison commented:
I opened:
- https://issues.apache.org/jira/browse/KAFKA-13258
- https://issues.apache.org/jira/browse/KAFKA-13259
- https://issues.apache.org/jira/browse/KAFKA-13260

We can't easily split the PR as the test makes them all fail

dajac commented:
Thanks. I wonder if we should consider them as blockers for 3.0. KAFKA-13258 is clearly a regression introduced in 2.8 and KAFKA-13260 could be quite bad as well. What do you think?

mimaison commented:
KAFKA-13260 is a regression in 3.0. However, I'm not sure if it actually has any impact. Looking at all callers of `errorCounts()`, it looks like it's mostly used to check for `NOT_CONTROLLER` error which can't happen with `FindCoordinator`. 

KAFKA-13258 is a regression in 2.8. This is directly visible by users when calling `Admin.alterClientQuotas()`, so I consider this high impact. 

KAFKA-13259 is here since the `DescribeProducers` API was introduced in 2.8. But `Admin.describeProducers()` is only being added in 3.0. I've not tried explicitly with a 3.0 client but I expect it behaves like KAFKA-13258, ie you don't get an Exception even when the call fails, so it's also high impact.

@kkonstantine can you take a look an decide whether we want this PR in 3.0? KAFKA-13258 and KAFKA-13259 are not technically regressions in 3.0 but are pretty annoying for users.

hachikuji commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -706,6 +706,8 @@ private void checkDescribeConfigsResponseVersions() {
         AbstractResponse response = req.getErrorResponse(e);
         checkResponse(response, req.version(), checkEqualityAndHashCode);
+        Map<Errors, Integer> errorCounts = response.errorCounts();
+        assertTrue(errorCounts.containsKey(Errors.forException(e)), "API Key " + req.apiKey().name + "V" + req.version() + " failed errorCounts test");
```
Is it possible to make this assertion stronger? Would we expect any other errors in the response?

mimaison replied:
@hachikuji Thanks for the review! I pushed an update

hachikuji reviewed (commented):
@mimaison Good finds and thanks for the patch. I think it's worth creating JIRAs for these issues since they affect previous releases. Really there's three bugs here. We can address them all here though since they are small.

hachikuji commented:
@mimaison Ah, I missed you already opened JIRAs. Never mind my comment.

mimaison committed:
Update test

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeProducersRequest.java:
```diff
@@ -69,7 +71,7 @@ public DescribeProducersRequestData data() {
     public DescribeProducersResponse getErrorResponse(int throttleTimeMs, Throwable e) {
         Errors error = Errors.forException(e);
-        DescribeProducersResponseData response = new DescribeProducersResponseData();
+        List<TopicResponse> topics = new ArrayList<>();
```
nit: We could keep instantiating `DescribeProducersResponseData` here and add the topic to it with `reponse.topics.add(..)`.

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/AlterClientQuotasRequest.java:
```diff
@@ -125,7 +127,10 @@ public AlterClientQuotasResponse getErrorResponse(int throttleTimeMs, Throwable
             }
-            responseEntries.add(new AlterClientQuotasResponseData.EntryData().setEntity(responseEntities));
+            responseEntries.add(new AlterClientQuotasResponseData.EntryData()
+                    .setEntity(responseEntities)
+                    .setErrorCode(error.code())
+                    .setErrorMessage(error.message()));
```
nit: Should we use 4 spaces to indent those ones here? We use 4 spaces for similar code at L127/128 so the code would remain a bit more homogeneous.

dajac reviewed (commented):
Thanks for the PR. I left a few nits. Sometimes, I feel like that we should add proper unit tests for all these methods. We've got bitten by similar issues few times now. At least, the one that you added should help to catch them in the future.

mimaison committed:
Address feedback

mimaison commented:
Thanks @dajac for the feedback. I've addressed your comments.

I agree with your comment about testing of these classes. `RequestResponseTest` is messy and it's hard to tell what's actually tested. I wouldn't be surprized if some requests or responses are completely missed!

hachikuji commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -706,6 +706,11 @@ private void checkDescribeConfigsResponseVersions() {
         checkResponse(response, req.version(), checkEqualityAndHashCode);
+        Errors error = Errors.forException(e);
+        Map<Errors, Integer> errorCounts = response.errorCounts();
+        assertEquals(1, errorCounts.size());
```
nit: could probably simplify these two assertions a little:
```java
assertEquals(Collections.singleton(error), errorCounts.keySet());
```

hachikuji reviewed (approved):
LGTM. Left a minor suggestion.

dajac reviewed (approved):
LGTM. Could we backport it to 3.0 and 2.8?

mimaison committed:
Tidy up assertions

mimaison commented:
Backports:
- 2.8: https://github.com/apache/kafka/pull/11299
- 3.0: https://github.com/apache/kafka/pull/11300

showuon commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeProducersRequest.java:
```diff
@@ -80,6 +80,7 @@ public DescribeProducersResponse getErrorResponse(int throttleTimeMs, Throwable
                         .setErrorCode(error.code())
```
should we add `.setErrorMessage(error.message())` here?

showuon reviewed (commented):
Great catch! Left a comment. Thank you.

showuon commented:
Also, this PR title should be updated to link to JIRA tickets. Thanks.
